src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp

```diff
@@ -573,35 +597,55 @@ void ParquetBlockInputFormat::decodeOneChunk(size_t row_group_batch_idx, std::un
+                .row_group_batch_idx = row_group_batch_idx,
+                .approx_original_chunk_size = approx_chunk_original_size
+        };
+    };
 
-    if (!row_group_batch.record_batch_reader)
+    if (!row_group_batch.record_batch_reader && !row_group_batch.native_record_reader)
         initializeRowGroupBatchReader(row_group_batch_idx);
 
-    auto batch = row_group_batch.record_batch_reader->Next();
-    if (!batch.ok())
-        throw Exception(ErrorCodes::CANNOT_READ_ALL_DATA, "Error while reading Parquet data: {}", batch.status().ToString());
-
-    if (!*batch)
+    PendingChunk res;
+    if (format_settings.parquet.use_native_reader)
     {
-        end_of_row_group();
-        return;
+        auto chunk = row_group_batch.native_record_reader->readChunk();
+        if (!chunk)
+        {
+            end_of_row_group();
+            return;
+        }
+
+        auto num_rows = chunk.getNumRows();
+        res = get_pending_chunk(num_rows, std::move(chunk));
     }
+    else
+    {
+        auto batch = row_group_batch.record_batch_reader->Next();
+        if (!batch.ok())
+            throw Exception(ErrorCodes::CANNOT_READ_ALL_DATA, "Error while reading Parquet data: {}", batch.status().ToString());
 
-    auto tmp_table = arrow::Table::FromRecordBatches({*batch});
+        if (!*batch)
+        {
+            end_of_row_group();
+            return;
+        }
 
-    size_t approx_chunk_original_size = static_cast<size_t>(std::ceil(static_cast<double>(row_group_batch.total_bytes_compressed) / row_group_batch.total_rows * (*tmp_table)->num_rows()));
-    PendingChunk res = {
-            .chunk = {},
-            .block_missing_values = {},
-            .chunk_idx = row_group_batch.next_chunk_idx,
-            .row_group_batch_idx = row_group_batch_idx,
-            .approx_original_chunk_size = approx_chunk_original_size
-    };
+        auto tmp_table = arrow::Table::FromRecordBatches({*batch});
+        res = get_pending_chunk((*tmp_table)->num_rows());
 
-    /// If defaults_for_omitted_fields is true, calculate the default values from default expression for omitted fields.
-    /// Otherwise fill the missing columns with zero values of its type.
-    BlockMissingValues * block_missing_values_ptr = format_settings.defaults_for_omitted_fields ? &res.block_missing_values : nullptr;
-    row_group_batch.arrow_column_to_ch_column->arrowTableToCHChunk(res.chunk, *tmp_table, (*tmp_table)->num_rows(), block_missing_values_ptr);
+        /// If defaults_for_omitted_fields is true, calculate the default values from default expression for omitted fields.
+        /// Otherwise fill the missing columns with zero values of its type.
+        BlockMissingValues * block_missing_values_ptr = format_settings.defaults_for_omitted_fields ? &res.block_missing_values : nullptr;
+        row_group_batch.arrow_column_to_ch_column->arrowTableToCHChunk(res.chunk, *tmp_table, (*tmp_table)->num_rows(), block_missing_values_ptr);
```

Comment:
Sorry, I missed this feature. I would like to add a TODO comment, because the block-missing-values logic will be modified again when supporting nested type, so the it can be implemented later for all types.